### Part 1 - Creating and evaluating an initial model

#### Code (15 marks)

In [182]:
## Importing the libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
data= pd.read_csv("rent_1.csv")

In [183]:
# Randomly looking at the values of the dataset
data.sample(10)

Unnamed: 0  bathrooms  bedrooms                       building_id  \
9037         9037        2.0         4  aed67db466854ec2aa5d249a032bc401   
15867       15867        1.5         2  a1bd0631f4febfb498e34b7cf68bcfb1   
14509       14509        1.0         2  6ce872b483cfcbb32ea805604d44ef5f   
2854         2854        1.0         3  d89f514c3ed0abaae52cba7017ac0701   
13388       13388        1.0         1  7a29be0848a05e5529631f4cfb2a830e   
15938       15938        2.0         2  c40e3e74475a91aae1928541be8df678   
17154       17154        1.0         2  3a956bd42c50f06ac84cf072fc514f5f   
4608         4608        1.0         2  f4447c422be55a383be80060a1f8685e   
13551       13551        1.5         4  af611427cf4c25dfcd281196154b267a   
4049         4049        2.0         2  2617d3212a513723a1518bff2caf0a33   

                   created                                        description  \
9037   2016-04-22 05:41:18  Located in the heart of Midtown East?s Turtle ...   
15867  2016-04-16 02:45:21  Gorgeous and HUGE 2 Bedroom apartment availabl...   
14509  2016-06-07 03:08:13  Beautifully renovated convertible 2 bedroom wi...   
2854   2016-06-24 08:22:18  ***PRIME MIDTOWN EAST OFF PARK AVE***TRUE 3 BE...   
13388  2016-05-05 05:57:12  Brand new and ON Central Park West!-Full servi...   
15938  2016-05-03 02:11:23  This delightful 2-bedroom, 2-bath apartment is...   
17154  2016-04-15 06:27:07  RENOVATED & BEAUTIFUL TRUE 2 bedroom (NOT A FL...   
4608   2016-04-30 03:19:52  Spacious 1 bedroom,-- Must See * ****** For mo...   
13551  2016-05-22 03:12:10  This beautiful loft is sure to bring out the A...   
4049   2016-04-02 02:17:22  Modern 2 bedroom/ 2 bath apartment featuring f...   

         display_address                                           features  \
9037    East 47th Street  ['Doorman', 'Elevator', 'Fitness Center', 'Pre...   
15867         E 33rd St.  ['Balcony', 'Doorman', 'Elevator', 'Laundry in...   
14509         Spruce St.  ['Laundry in Building', 'Laundry in Unit', 'Di...   
2854           E 54th St  ['Dining Room', 'Elevator', 'Laundry in Buildi...   
13388  Central Park West  ['Dining Room', 'Doorman', 'Elevator', 'Dishwa...   
15938   West 96th Street  ['Doorman', 'Elevator', 'Outdoor Space', 'Dogs...   
17154         W 42nd St.  ['Swimming Pool', 'Roof Deck', 'Dining Room', ...   
4608         E 49 Street                                ['Hardwood Floors']   
13551        74 Franklin  ['Dining Room', 'Loft', 'Laundry in Building',...   
4049    West 52nd Street  ['Doorman', 'Elevator', 'Laundry in Unit', 'Ca...   

       latitude  longitude                        manager_id  \
9037    40.7530   -73.9695  c224f0c13e8bb14e79409501911fce73   
15867   40.7447   -73.9782  0319c8009f00e29f3d434c1dd1d25e74   
14509   40.7111   -74.0055  9ebb4938732754c8d5f32822ff0219da   
2854    40.7594   -73.9712  e90f2ded843cdb2efd65ef47d9fc8029   
13388   40.7843   -73.9703  2777b4cbbabcaecd4c14b703de27a668   
15938   40.7959   -73.9742  487d6755e992da0fb090cf7d79298f1c   
17154   40.7613   -73.9998  1fb46c4a72bcf764ac35fc23f394760d   
4608    40.7537   -73.9682  41735645e0f8f13993c42894023f8e58   
13551   40.7180   -74.0042  0ca47ea1b8b87069bc53d3fcbce19019   
4049    40.7664   -73.9918  8b1ca4681cd1d4b8c7392e7ccbedd052   

                                                  photos  price  \
9037   ['https://photos.renthop.com/2/6913120_44e385b...   6500   
15867  ['https://photos.renthop.com/2/6882156_09b9bef...   4195   
14509  ['https://photos.renthop.com/2/7118805_86e40d7...   3750   
2854   ['https://photos.renthop.com/2/7211689_79e2338...   4350   
13388  ['https://photos.renthop.com/2/6970057_5995635...   4850   
15938  ['https://photos.renthop.com/2/6955156_1d414ca...   4750   
17154  ['https://photos.renthop.com/2/6879237_c4c083b...   3300   
4608   ['https://photos.renthop.com/2/6946188_0675a04...   2575   
13551  ['https://photos.renthop.com/2/7052457_3fd5186...   9600   
4049   ['https://photos.renthop.c

In [184]:
type(data)

pandas.core.frame.DataFrame

In [185]:
# Saving only the integer and float values in the dataset in the variable named "data"

data= pd.read_csv("rent_1.csv",usecols=['bathrooms','bedrooms','latitude','longitude','interest_level','price'])

In [186]:
# Coverting the dataset into dataframe.

df = pd.DataFrame(data)

In [187]:
# Looking at the top 5 rows of the dataframe

df.head()

bathrooms  bedrooms  latitude  longitude  price  interest_level
0        1.0         2   40.7272   -73.9829   4145               1
1        1.0         1   40.7471   -74.0023   4425               1
2        1.0         1   40.7738   -73.9595   3200               1
3        1.0         1   40.7586   -73.9838   2700               1
4        1.0         1   40.7522   -73.8830   1850               2

### Now, whole dataset will be split into two parts:
- Training part : For our model training.
- Test part : For model evaluation.

In [188]:
## Assigning the training data in the variable called 'X'. Since, price needs to be predicted so we will not keep that in X.

X = df.drop(['price'],axis=1)

In [189]:
## Assigning price in variable 'Y' to train the model so that it predict correct price based on the data given in X variale.

y= df['price']

In [190]:
X.head()

bathrooms  bedrooms  latitude  longitude  interest_level
0        1.0         2   40.7272   -73.9829               1
1        1.0         1   40.7471   -74.0023               1
2        1.0         1   40.7738   -73.9595               1
3        1.0         1   40.7586   -73.9838               1
4        1.0         1   40.7522   -73.8830               2

In [191]:
y.head()

0    4145
1    4425
2    3200
3    2700
4    1850
Name: price, dtype: int64

In [192]:
## Importing train_test split library from sklearn.model package
## Splitting X and Y as train and test with test_size= 30%

from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test= train_test_split(X,y,test_size=0.3,random_state=42)

In [193]:
## Checking how many rows and columns have randomly been taken for training and testing the model. Since, Y has only 1 column, price
## for training and testing therefore, it only shows the number of rows in the shape.

print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(14000, 5)
(6000, 5)
(14000,)
(6000,)


In [194]:
## Now, fit the Decision tree Regressor on our training set.

In [195]:
## Importing Random Forest Regressor

from sklearn.ensemble import RandomForestRegressor

In [196]:
## Training model using random forest regressor. Where, number of decision trees would be 200 and as we want OOB score so we have done True for it.

rfr = RandomForestRegressor(n_estimators=200,random_state=0,oob_score = True)

In [197]:
rfr.fit(X_train,y_train)

RandomForestRegressor(n_estimators=200, oob_score=True, random_state=0)

In [198]:
# Evaluate the model by using performance metric (R2-Score)

In [199]:
from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error

In [200]:
## Calculating R2 score and Mean Absolute Error for the models we have tested after giving training

print("R2 Score is:",np.round(r2_score(y_test,rfr.predict(X_test))*100,2))
print("MAE:",mean_absolute_error(y_test,rfr.predict(X_test).round(3)))

R2 Score is: 69.87
MAE: 525.4922253333333


In [201]:
print('Score: ', (rfr.score(X_train, y_train)*100).round(2))

Score:  78.75


In [202]:
print('OOB Score:',(rfr.oob_score_*100).round(3))

OOB Score: -40.133


### We can see that OOB sore is negative that means our model is worse than the best constant predictor.


- The overall objective is to predict the price based on the number of features in the dataset. However, after seeing the dataset few features were removed because that features do not add values. Moreover, our algorithm only understand numerical features for prediction.

- Next, RandomforestRegressor has been applied on training set without any data mining/cleaning and different scores calculated by using different metrics.

- Model is giving ~70% accuracy score and R2 score but OOB score is negative(~40%) that means our model is perfroming very poor on unseen data.

- In order to make this better hyperparameter tunning and data exploration.cleaning is required that we are going to perform further.



### Part 2 - Denoising the data

This section should include the code necessary to **denoise** the data, and it should include what is needed to identify inconsistencies, problems, errors, etc. in the data. 


In [230]:
## Loading the uncleaned data again with all the columns and rows

data= pd.read_csv("rent_1.csv")

In [231]:
data.head()

Unnamed: 0  bathrooms  bedrooms                       building_id  \
0           0        1.0         2  c1ca7a55a2b7f478b580940f5982e54e   
1           1        1.0         1  7183d18cb8f135986bae1536396b3e23   
2           2        1.0         1  195629fb6278370e4a127427cb92c41e   
3           3        1.0         1                                 0   
4           4        1.0         1  0c622d4d2688e39f5cec16dfea1c16f6   

               created                                        description  \
0  2016-06-01 03:14:01  Actual Apartment Photos`Spacious Studio Apartm...   
1  2016-04-02 02:32:53  CooperCooper.com :: Web ID #051243; Access 100...   
2  2016-05-18 02:22:07  Beautiful LARGE one bedroom apartment - Great ...   
3  2016-04-24 01:10:38  Great one bedroom in prime Midtown location! W...   
4  2016-05-06 02:13:10  This is a super spacious one bedroom in Jackso...   

    display_address                                           features  \
0        E 9 Street  ['Dishwasher', 'Hardwood Floors', 'Cats Allowed']   
1  West 23rd Street  ['Doorman', 'Fitness Center', 'Pool', 'Elevato...   
2  Lexington Avenue  ['Furnished', 'Hardwood Floors', 'Dogs Allowed...   
3  West 47th Street                                        ['Doorman']   
4          35th Ave                   ['Cats Allowed', 'Dogs Allowed']   

   latitude  longitude                        manager_id  \
0   40.7272   -73.9829  23fe9db389d31048a104bd4686879a77   
1   40.7471   -74.0023  6e5c10246156ae5bdcd9b487ca99d96a   
2   40.7738   -73.9595  c71cf1f472cf9b4b4517ea23fb6f2c91   
3   40.7586   -73.9838  048ff8c9b8a4831e117e113d3cc5926b   
4   40.7522   -73.8830  8cdddf7fdb66d4db900fe5b5b62cbc47   

                                              photos  price  \
0  ['https://photos.renthop.com/2/7089465_2987f65...   4145   
1  ['https://photos.renthop.com/2/6813513_db8cbea...   4425   
2  ['https://photos.renthop.com/2/7027889_fbea270...   3200   
3                                                 []   2700   
4  ['https://photos.renthop.com/2/6972708_fb08bfd...   1850   

          street_address  interest_level  
0         442 E 9 Street               1  
1   425 West 23rd Street               1  
2  1109 Lexington Avenue               1  
3   150 West 47th Street               1  
4         83-09 35th Ave               2

In [232]:
## Removing unwanted columns to clean the data

clean_data= data.drop(['building_id', 'created', 'Unnamed: 0', 'description', 'display_address', 'manager_id', 'photos', 'street_address'],axis=1)
clean_data

bathrooms  bedrooms                                           features  \
0            1.0         2  ['Dishwasher', 'Hardwood Floors', 'Cats Allowed']   
1            1.0         1  ['Doorman', 'Fitness Center', 'Pool', 'Elevato...   
2            1.0         1  ['Furnished', 'Hardwood Floors', 'Dogs Allowed...   
3            1.0         1                                        ['Doorman']   
4            1.0         1                   ['Cats Allowed', 'Dogs Allowed']   
...          ...       ...                                                ...   
19995        1.0         1  ['Doorman', 'Pre-War', 'Dogs Allowed', 'Cats A...   
19996        2.0         2  ['Dining Room', 'Doorman', 'Fitness Center', '...   
19997        1.0         3                                                 []   
19998        1.0         0        ['Pre-War', 'Dogs Allowed', 'Cats Allowed']   
19999        1.0         2                              ['Pre-War', 'No Fee']   

       latitude  longitude  price  interest_level  
0       40.7272   -73.9829   4145               1  
1       40.7471   -74.0023   4425               1  
2       40.7738   -73.9595   3200               1  
3       40.7586   -73.9838   2700               1  
4       40.7522   -73.8830   1850               2  
...         ...        ...    ...             ...  
19995   40.7468   -73.9818   2900               1  
19996   40.7787   -73.9826   9000               1  
19997   40.7650   -73.9882   2800               2  
19998   40.7808   -73.9766   2100               1  
19999   40.8018   -73.9654   3000               1  

[20000 rows x 7 columns]

In [233]:
## Now, let's see if there are any null values in the dataset

clean_data.isnull().sum()

bathrooms         0
bedrooms          0
features          0
latitude          0
longitude         0
price             0
interest_level    0
dtype: int64

In [234]:
a = clean_data[clean_data['price'] <= 0]

In [235]:
## Deleting the rows where price is either less than 0 or equal to 0

clean_data = clean_data.drop(a.index, axis=0)

In [236]:
## Counting number of features in each house and giving the count corresponding to it. Eg: 1st have 3 features, 2nd has 11.

clean_data['features']=clean_data['features'].apply(lambda x : len(x[1:-1].split(",")))

In [237]:
clean_data['features']

0         3
1        11
2         4
3         1
4         2
         ..
19995     4
19996     7
19997     1
19998     3
19999     2
Name: features, Length: 19979, dtype: int64

In [239]:
# Replacing blank values with mean i.e. 5

clean_data['features']=clean_data['features'].replace(to_replace='[]',value=5)

In [240]:
## Checking if there is still any blank value left in the dataset

clean_data[clean_data['features'] == '[]'].count()

bathrooms         0
bedrooms          0
features          0
latitude          0
longitude         0
price             0
interest_level    0
dtype: int64

In [241]:
clean_data.head()

bathrooms  bedrooms  features  latitude  longitude  price  interest_level
0        1.0         2         3   40.7272   -73.9829   4145               1
1        1.0         1        11   40.7471   -74.0023   4425               1
2        1.0         1         4   40.7738   -73.9595   3200               1
3        1.0         1         1   40.7586   -73.9838   2700               1
4        1.0         1         2   40.7522   -73.8830   1850               2

In [242]:
clean_data.describe()

bathrooms      bedrooms      features      latitude     longitude  \
count  19979.000000  19979.000000  19979.000000  19979.000000  19979.000000   
mean       1.210396      1.535662      5.500075     40.697729    -73.824288   
std        0.504403      1.115812      3.875443      1.777423      4.379368   
min        0.000000      0.000000      1.000000    -17.644000    -93.270500   
25%        1.000000      1.000000      2.000000     40.728300    -73.991900   
50%        1.000000      1.000000      5.000000     40.751500    -73.978100   
75%        1.000000      2.000000      8.000000     40.774400    -73.954800   
max        6.500000      8.000000     36.000000     44.883500     77.181000   

              price  interest_level  
count  1.997900e+04    19979.000000  
mean   3.919868e+03        1.382952  
std    3.185228e+04        0.646073  
min    4.500000e+01       -3.000000  
25%    2.495000e+03        1.000000  
50%    3.150000e+03        1.000000  
75%    4.100000e+03        2.000000  
max    4.490000e+06        3.000000

In [243]:
clean_data.head()

bathrooms  bedrooms  features  latitude  longitude  price  interest_level
0        1.0         2         3   40.7272   -73.9829   4145               1
1        1.0         1        11   40.7471   -74.0023   4425               1
2        1.0         1         4   40.7738   -73.9595   3200               1
3        1.0         1         1   40.7586   -73.9838   2700               1
4        1.0         1         2   40.7522   -73.8830   1850               2

In [244]:
## Considering only those where price lies between 1000 to 10000

clean_data = clean_data[(clean_data['price'] > 1000) & (clean_data['price'] < 10000)]

In [245]:
clean_data.shape

(19538, 7)

In [246]:
clean_data[clean_data['latitude'] < 40]

bathrooms  bedrooms  features  latitude  longitude  price  \
93           1.0         1         8  -17.6440    77.1810   3645   
1433         1.0         1         7    0.0000     0.0000   3600   
4994         2.5         4        12  -17.6440    77.1810   5995   
5296         1.0         1         4  -17.6440    77.1810   3300   
5562         1.0         2        10  -17.6440    77.1810   3150   
5754         1.0         0         1    0.0000     0.0000   1850   
6243         1.0         0         4  -17.6440    77.1810   2450   
8171         1.0         1         3  -17.6440    77.1810   2300   
9598         1.0         1         2    0.0000     0.0000   3495   
11724        1.0         0         3  -17.6440    77.1810   1975   
12415        1.0         2         4    0.0000     0.0000   3200   
13811        1.0         0         1  -17.6440    77.1810   1950   
14221        1.0         1         3  -17.6440    77.1810   4065   
14247        1.0         2         9  -17.6440    77.1810   1900   
14719        1.0         0         7   39.8395   -86.1527   2650   
15043        1.0         2        14  -17.6440    77.1810   3400   
16174        1.0         1         3  -17.6440    77.1810   1750   
16231        5.0         6         6    0.0000     0.0000   9995   
16670        1.0         2         8  -17.6440    77.1810   4400   
17255        1.0         2         1  -17.6440    77.1810   2500   
17506        1.0         2         3    0.0000     0.0000   2950   
17746        1.0         1         3    0.0000     0.0000   1750   
19159        1.0         3         1  -17.6440    77.1810   2750   

       interest_level  
93                  1  
1433                1  
4994                1  
5296                1  
5562                1  
5754                1  
6243                1  
8171                2  
9598                1  
11724               1  
12415               1  
13811               2  
14221               1  
14247               2  
14719               1  
15043               1  
16174               3  
16231               1  
16670               1  
17255               2  
17506               2  
17746               1  
19159               2

In [247]:
 l = clean_data[clean_data['latitude'] ==0]

In [248]:
l.index

Int64Index([1433, 5754, 9598, 12415, 16231, 17506, 17746], dtype='int64')

In [249]:
## Dropping those rows where latitue is 0 because they are not useful to us

clean_data=clean_data.drop(l.index,axis=0)

In [250]:
clean_data = clean_data[(clean_data['latitude'] !=0) | (clean_data['longitude']) ]

In [251]:
clean_data.shape

(19531, 7)

### Part 3 - Creating and evaluating a final model


In [252]:
X = clean_data.drop(['price'],axis=1)
X.head()

bathrooms  bedrooms  features  latitude  longitude  interest_level
0        1.0         2         3   40.7272   -73.9829               1
1        1.0         1        11   40.7471   -74.0023               1
2        1.0         1         4   40.7738   -73.9595               1
3        1.0         1         1   40.7586   -73.9838               1
4        1.0         1         2   40.7522   -73.8830               2

In [253]:
y=clean_data['price']

In [254]:
y.head()

0    4145
1    4425
2    3200
3    2700
4    1850
Name: price, dtype: int64

In [255]:
## Training the model and testing with size 30%

X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.3,random_state=132)

In [256]:
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(13671, 6)
(5860, 6)
(13671,)
(5860,)


In [257]:
rfr1 = RandomForestRegressor(n_estimators=200,oob_score=True)

In [258]:
## Training the model

rfr1.fit(X_train,y_train)

RandomForestRegressor(n_estimators=200, oob_score=True)

In [264]:
## calculating R squared score

print("R2_Score1:",r2_score(y_test,rfr1.predict(X_test))*100)

R2_Score1: 82.45680576205638


In [260]:
## calculating Mean Absolute Error

print("MAE1:",mean_absolute_error(y_test,rfr1.predict(X_test).round(3)))

MAE1: 378.30848839590436


In [263]:
## calculating Out-Of-Bag (OOB) score

print('OOB Score1:',(rfr1.oob_score_*100).round(3))

OOB Score1: 80.137


#### Explanation (5 marks)

Please provide an explanation and justification for the code submitted in **Part 3** in the context of the overall objective. 

### Answer

Since, our algorithm only understand numerical features for prediction so I first encoded the features columns into numerical by 
calculating count of features for each house using lambda function.

Next, RandomforestRegressor has been applied on training set after cleaning the data and removing few rows and columns. Post that different scores are calculated such as R squared, Mean Absolute Error and Out-Of-Bag.

Model is giving 82% as R2 score, more the R2 score better it is. Furthermore, OOB score is increased from negative 40 to positive 80 that means our model is perfroming very good on unseen data.

The above scores tells how important the cleaning is for improving the performance of our model.

### Issues faced during working on the dataset

| Data issue discovered    | Why is this a problem?    | How did you fix it?    | Why is this fix appropriate?   |
| :- | :- | :- | :- | 

| **Converting features into integer**   |  I wanted to use features for training the model because more the features higher the price. Since machines only understand integers, floats so I had to somehow convert it into appropriate data type   | Using 'Lembda' function I counted how many feature does each house has and wrote the value corresponding to it. Eg: if any house has 3 differnt features I mentioned 3 and for 4 I mentioned 4.   | Features can be considered as an important part to predict the price of houses. Also, in the dataset we can observe that highest featured house has highest price.

| **Building ID**    | Building_id is unnecessary and shall be removed    | I dropped the column named 'building_id'    | We should only keep important and those features which can be used to train the model.

| **Created**   | Created column is not useful in this dataset   | I dropped this column to make the data look clean and sorted | To keep the data clean we should remove all columns and rows that are not useful.

| **Unnamed**   | As indexing is automatically been done for all rows so we technically do not require Unnamed to give serial number to each row  | Removed this column  | It was creating confusion so after removing this column our dataset became more understandable and clear

| **Description**  | Model uses only integer and float values to train itself. Therefore, there is no need of keeping string column which is not useful  | If we have only useful data, our model will be trained better and will not get confused

| **Display_Address**   | This column is also not useful to us hence shall be removed   | I dropped this column   | We should only keep useful data for better model training

| **Manager_id**   | Created column is not useful in this dataset  | Removed manager_id column  | We should only keep useful data for better model training

| **Photos**  | Photos column was not useful for training the model  | Dropped the column Photos  | We should only keep useful data for better model training

| **Street_address** | There is no use of this column | I thus, removed this column  | | We should only keep useful data for better model training

| **Blank values in Feature column**   | There were 1341 blank values in features' column  | I replaced these blank cells with mean value i.e. 5   | Blank cells confuses the model and affects the accuracy of the model. Therefore, replaced it with best suitable value i.e. mean of that whole column

| **Rent Price**   | I figured out there were few values that was less than 1000 and more than 10000. These can be considered as outliers and are not correct  | I removed few rows where price < 1000 and price > 10000  | Outliers can affect the accuracy of the model and thus shall be removed

| **0 degree Latitude**  | There have been few rows where latitude was 0 degree which are not useful here as New York is not at equator  | I removed those rows  | It helped keeping only useful data and model will be able to only use accurate data for training itself


Prior cleaning the above dataset our
- **R2 Score** is: 69.87
- **MSE:** 525.4922253333333 
- **OOB score** is: -40.133 

-- This tells there were so many errors and R2 Score is too high
-- MAE also is high and should be reduced. On the other side, our OOB is -40 which is too bad that means our model is worse than the best constant predictor.


-- Whereas, after cleaning the data and feature engineering and selection resolved the issues and our model performance increased drastically.
- **R2 Score** is 82 which is very good number.
- **MAE** is 377 which is also reduced and got better than before.
- **OOB Score**: 80.102 which can be said as very good.